### PR TITLE
Engine: sync contract id internal representation with spec

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR, we inline the internal representation of absolute contract id with the daml-lf spec. 

This PR advances the state of #3830.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 1]
```diff
@@ -347,9 +347,66 @@ object Value extends CidContainer1WithDefaultCidResolver[Value] {
...
+      def fromString(s: String): Either[String, V0] =
+        Ref.ContractIdString.fromString(s).map(V0(_))
+
+      def assertFromString(s: String): V0 = assertRight(fromString(s))
+    }
+
+    final case class V1(discriminator: crypto.Hash, suffix: Bytes = Bytes.Empty)
+        extends AbsoluteContractId {
+      lazy val toBytes: Bytes = V1.prefix ++ discriminator.bytes ++ suffix
+      lazy val coid: Ref.HexString = toBytes.toHexString
```
Comment (gerolf-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: Often a contractId is bound to a name like `coid`. Referencing the hex string the looks like `coid.coid`. How about we call this `toHexString`? Also makes it very explicit what this is.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: :+1:

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: We need the same name for V0 and V1. V0 coid is not a `HexString`.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: I definitively do not want to use `toString`. Any suggestion ?

Comment (gerolf-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala]: okay, let's keep it for now.

**Pull request merged**